esmtools/stats.py

```diff
@@ -135,317 +132,381 @@ def area_weight(da, area_coord="area"):
-    this might be altered in the future.
-    Parameters
-    ----------
-    da : xarray DataArray
-    dim : str
-        dimension to smooth over (e.g. 'time')
-    length : int
-        number of steps to smooth over for the given dim
-    center : boolean (default to True)
-        whether to center the smoothing filter or start from the beginning
-    Returns
-    -------
-    smoothed : smoothed DataArray object
+def _check_y_not_independent_variable(y, dim):
+    """Checks that `y` is not the independent variable in statistics functions.
+
+    Args:
+        y (xr.DataArray or xr.Dataset): Dependent variable from statistics functions.
+        dim (str): Dimension statistical function is being applied over.
+
+    Raises:
+        ValueError: If `y` is a DataArray and equal to `dim`. This infers that something
+            like a time axis is being placed in the dependent variable.
     """
-    return da.rolling({dim: length}, center=center).mean()
+    if isinstance(y, xr.DataArray) and (y.name == dim):
+        raise ValueError(
+            f"Dependent variable y should not be the same as the dim {dim} being "
+            "applied over. Change your y variable to x."
+        )
 
 
-@is_xarray(0)
-def fit_poly(ds, order, dim="time"):
-    """Returns the fitted polynomial line of order N
+def _convert_time_and_return_slope_factor(x, dim):
+    """Converts `x` to numeric time (if datetime) and returns slope factor.
 
-    .. note::
-        This automatically interpolates across NaNs to make the fit.
+    This accounts for differences in length of months, leap years, etc. when fitting
+    a regression and also ensures that the numpy functions don't break with datetimes.
 
     Args:
-        ds (xarray object): Time series to fit  polynomial to.
-        order (int): Order of polynomial fit.
-        dim (optional str): Dimension over which to fit hte polynomial.
+        x (xr.DataArray or xr.Dataset): Independent variable from statistical functions.
+        dim (str): Dimension statistical function is being applied over.
 
     Returns:
-        xarray object with polynomial fit.
-
-    References:
-        This is a modification of @ahuang11's script `rm_poly` in `climpred`.
+        x (xr.DataArray or xr.Dataset): If `x` is a time axis, converts to numeric
+            time.
+        slope_factor (float): Factor to multiply slope by if returning regression
+            results. This accounts for the fact that datetimes are converted to
+            "days since 1990-01-01" numeric time and thus the answer comes out
+            in the original units per day (e.g., degC/day). This auto-converts to
+            the original temporal frequency (e.g., degC/year) if the calendar
+            can be inferred.
     """
-    has_dims(ds, dim, "dataset")
+    slope_factor = 1.0
+    if isinstance(x, xr.DataArray):
+        x_type = x.timeutils.type
+        if x_type in ["DatetimeIndex", "CFTimeIndex"]:
```

Comment:
Correct. Because it could just be integer time steps/years which we can't decipher. And integer years are close enough in size (365 vs 366) to not yield much difference in slope anyways. Also cases for like Nino indices that aren't time.